waspc/data/Generator/templates/server/src/auth/providers/config/github.ts

```diff
@@ -0,0 +1,149 @@
+import { getStateCookieName, getValueFromCookie, setValueInCookie } from "../oauth/cookies.js";
+import { tokenStore } from "../oauth/oneTimeCode.js";
+import { callbackPath, getRedirectUriForOneTimeCode } from "../oauth/redirect.js";
+
+{=# userSignupFields.isDefined =}
+{=& userSignupFields.importStatement =}
+const _waspUserSignupFields = {= userSignupFields.importIdentifier =}
+{=/ userSignupFields.isDefined =}
+{=^ userSignupFields.isDefined =}
+const _waspUserSignupFields = undefined
+{=/ userSignupFields.isDefined =}
+{=# configFn.isDefined =}
+{=& configFn.importStatement =}
+const _waspUserDefinedConfigFn = {= configFn.importIdentifier =}
+{=/ configFn.isDefined =}
+{=^ configFn.isDefined =}
+const _waspUserDefinedConfigFn = undefined
+{=/ configFn.isDefined =}
+
+/*
+
+Consider introducing a helper called
+
+createRouter({
+    initAuthProvider: () => new GitHub(),
+    prepareRedirectUrl: async (req, res, provider, auth) => {},
+    getProviderProfile: async (req, res, provider, auth) => {},
+    getProviderUserId: (providerProfile) => {},
+})
+
+Or consider multilpe small helpers like
+
+setStateCookie(provider.id, state, res)
+getStateCookie(provider.id, req)
+getCodeFromQuery(req)
+validateState(storedState, state, code)
+validateCode(code)
+createAuthorizationUrl(provider, state)
+
+*/
+
+const _waspConfig: ProviderConfig = {
+    id: "{= providerId =}",
+    displayName: "{= displayName =}",
+    createRouter(provider) {
+        const router = Router();
+
+        // TODO: make sure to validate the env vars
+        const github = new GitHub(
+            process.env.GITHUB_CLIENT_ID,
+            process.env.GITHUB_CLIENT_SECRET,
+        );
+
+        router.get('/login', async (_req, res) => {
+            const state = generateState();
+            const url = await github.createAuthorizationURL(state, {
+                // TODO: use the user defined config function
+                scopes: {=& requiredScopes =},
+            });
+            setValueInCookie(getStateCookieName(provider.id), state, res);
+            return res.status(302).setHeader("Location", url.toString()).end();
+        })
+
+        router.get(`/${callbackPath}`, async (req, res) => {
+            const storedState = getValueFromCookie(
+                getStateCookieName(provider.id),
+                req
+            );
+            const state = req.query.state;
+            const code = req.query.code;
+    
+            if (
+                !storedState ||
+                !state ||
+                storedState !== state ||
+                typeof code !== "string"
+            ) {
+                return res.status(400).json({
+                    success: false,
+                    message: "Invalid state",
+                });
+            }
+    
+            try {
+                const { accessToken } = await github.validateAuthorizationCode(code);
+    
+                // TODO: maybe an additional request to get the user's email?
```

Comment:
The passport strategy did it by default if the scopes included `user` or `user:email` https://github.com/cfsghost/passport-github/blob/master/lib/strategy.js#L119

Reply:
What are the options? Either we get it for them, or they get it on their own.

Maybe let's do the same as Passport strategy did, it sounds useful in practice, and will be a less of a breaking change.

Reply:
Okay, this comment got outdated with latest changes. I replicated the logic that Passport had to keep the same information available. I think that makes the most sense!